Check the validity of a preimage

linked0 opened the pull request:
When new pre-image is revealed, it must be checked if it's the right pre-image for a current pre-image stored in the validator set. This PR adds the `isInvalidPreimageReason` function checking the validity of a preimage and makes the validator set use the function in the `addPreimage` function.

Relates #694

linked0 committed:
Add function checking validity of two pre-images

It checks if the pre-image with higher height is a pre-image of the pre-image
with lower height.

AndrejMitrovic commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,53 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+
+    Params:
+        pre_image = The pre-image information
+        cur_image = The current image information
```
How about naming them `prev_image` and then `next_image` or `new_image`, just to make it obvious which preimage was already in the EnrollmentManager?

linked0 replied:
> How about naming them prev_image and then next_image or new_image, just to make it obvious which preimage was already in the EnrollmentManager?

Good suggestion!

AndrejMitrovic commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,53 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+        return "the pre-image has a invalid height";
+
+    Hash temp_hash = pre_image.hash;
+    for (uint i = 0; i < pre_image.height - cur_image.height; i++)
```
What happens when a byzantine node sends a `PreimageInfo` with `pre_image.height` set to `ulong.max`?

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -526,6 +526,19 @@ public class ValidatorSet
+        // check the validity of new pre-image based on the stored pre-image
+        PreimageInfo cur_image;
+        if (this.getPreimage(preimage.enroll_key, cur_image) &&
+            cur_image != PreimageInfo.init)
```
Why would the preimage be default-initialized if `getPreimage` returns `true`? It seems to me that should never happen.

linked0 replied:
> Why would the preimage be default-initialized if getPreimage returns true? It seems to me that should never happen.

It's possible there is no pre-image. The return value of `getPreimage` is about whether the operation is processed succesfully, not about whether a pre-image exists.

linked0 commented:
Ready for review.
I applied the suggestion to the code and dealt with the problem related to the height of an preimage.

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+    const ref PreimageInfo prev_image, uint validator_cycle) nothrow @safe
+{
+    if (new_image.enroll_key != prev_image.enroll_key)
+        return "The pre-image has a wrong enrollment key";
```
```suggestion
        return "The pre-image's enrollment key differ from it's descendant";
```

Rationale for the wording: "wrong" doesn't mean much. It doesn't say what was expected, or how that expectation wasn't met.

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+    if (new_image.enroll_key != prev_image.enroll_key)
+        return "The pre-image has a wrong enrollment key";
+
+    if (new_image.height <= prev_image.height ||
```
Why the `<=` ? It could be just `<`, couldn't it ?

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+
+    if (new_image.height <= prev_image.height ||
+        new_image.height - prev_image.height >= validator_cycle)
+        return "the pre-image has a invalid height";
```
II would actually split this into two returns. One with `<` and one with the check against `validator_cycle` so you can provide a more useful error message.

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+        return "the pre-image has a invalid height";
+
+    Hash temp_hash = new_image.hash;
+    for (uint i = 0; i < new_image.height - prev_image.height; i++)
```
```suggestion
    foreach (_; prev_image.height .. new_image.height)
```

Not tested, but it *should* work, and it's much easier to see what is happening

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
+
+    Params:
+        new_image = The pre-image information to check
+        prev_image = The previous pre-image information
```
Missing `validator_cycle` documentation

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
 
+/*******************************************************************************
+
+    Check the validity of a pre-image information
```
You might want to express your documentation in terms of "parent" and "descendant" / "child".
Because there's a clear hierarchy in pre-image: we successively reveal "parent" or "ancestor" of a hash.

Documenting "pre-image" as invalid is a bit confusing to me, because a pre-image is not a self-contained entity, it's defined in terms of its image. What we can test is if a `PreimageInfo` is a parent for another `PreimageInfo`.

Consider the following scenario: You have a `PreimageInfo` at height 98, and you get a `PreimageInfo` for height 90. Except the hash actually matches the pre-image for height 92. The `PreimageInfo` is not a parent for the `PreimageInfo` we have, but the hash itself is a pre-image - just at a different height.

linked0 replied:
The `PreimageInfo` is a model for "pre-image" we described in the white paper. So we can use `PreimageInfo` and "pre-image" interchangingly. Because we are developing a "pre-image" in our product, we can use the word, "pre-image" as `PreimageInfo`. So the "pre-image" gets to have an actual preimage and a height. It's not the pre-image we use in Mathematics or Cryptography. We need to use an word like "pre-image" in our context.
I hope I understood what you mean exactly.

Geod24 replied:
> It's not the pre-image we use in Mathematics or Cryptography.

My point was that, since we are borrowing vocabulary from those field, we shouldn't conflate the two. Whenever talking about details, we could use `PreImageInfo`, but in the general sense we refer to them as "preimage".
But the documentation, as it stands, is good enough.

linked0 replied:
> My point was that, since we are borrowing vocabulary from those field, we shouldn't conflate the two. Whenever talking about details, we could use PreImageInfo, but in the general sense we refer to them as "preimage".

Thank your for the additional comment and the merge.

Geod24 reviewed (changes requested):
This need a test in `agora.test`

linked0 committed:
Check the validity of a pre-image

Using the isInvalidPreimageReason function, the validator set checks
the validity of the pre-image passed in the addPreimage function.

Geod24 reviewed (approved):
LGTM
